lib/active_merchant/billing/gateways/xpay.rb

```diff
@@ -30,77 +30,162 @@ def initialize(options = {})
+          post[:order][:customerInfo][:shippingAddress] = {
+            name: address[:name],
+            street: address[:address1],
+            additionalInfo: address[:address2],
+            city: address[:city],
+            postCode: address[:zip],
+            country: address[:country]
+          }.compact
+        end
+      end
+
+      def add_recurrence(post, options)
+        post[:recurrence] = { action: options[:recurrence] || 'NO_RECURRING' }
+      end
+
+      def add_exemptions(post, options)
+        post[:exemptions] = options[:exemptions] || 'NO_PREFERENCE'
+      end
+
+      def add_3ds_params(post, options)
+        post[:threeDSAuthData] = { threeDSAuthResponse: options[:three_ds_auth_response] }.compact
+        post[:operationId] = options[:operation_id] if options[:operation_id]
+      end
 
-      def add_payment_method(post, payment_method) end
+      def add_auth_purchase_params(post, amount, payment_method, options)
+        add_invoice(post, amount, options)
+        add_payment_method(post, payment_method)
+        add_customer_data(post, payment_method, options)
+        add_address(post, options)
+        add_recurrence(post, options) unless options[:operation_id]
+        add_exemptions(post, options)
+        add_3ds_params(post, options)
+      end
 
       def add_reference(post, authorization) end
 
       def add_auth_purchase(post, money, payment, options) end
 
+      def parse(body)
+        return {} if body.blank?
+
+        JSON.parse(body)
+      end
+
       def commit(action, params, options)
-        url = build_request_url(action)
-        response = ssl_post(url, params.to_json, request_headers(options))
-
-        unless response
-          Response.new(
-            success_from(response),
-            message_from(success_from(response), response),
-            response,
-            authorization: authorization_from(response),
-            avs_result: AVSResult.new(code: response['some_avs_result_key']),
-            cvv_result: CVVResult.new(response['some_cvv_result_key']),
-            test: test?,
-            error_code: error_code_from(response)
-          )
+        transaction_id = params.dig(:operation_id) unless action != 'capture'
+        begin
+          url = build_request_url(action, transaction_id)
+          raw_response = ssl_post(url, params.to_json, request_headers(options))
+          response = parse(raw_response)
+        rescue ResponseError => e
+          response = e.response.body
+          response = parse(response)
         end
-      rescue ResponseError => e
-        response = e.response.body
-        JSON.parse(response)
+
+        Response.new(
+          success_from(response),
+          message_from(response),
+          response,
+          authorization: authorization_from(response),
+          test: test?,
+          error_code: error_code_from(response)
+        )
       end
 
       def request_headers(options)
```

Comment:
💬 Style/ Comment:

Just to reduce unnecessary noise:

```Ruby
      def request_headers(options)
        {
          'Content-Type' => 'application/json',
          'X-Api-Key' => @api_key,
          'Correlation-Id' => options[:order_id] || SecureRandom.uuid
        }
      end

```

Reply:
added! thanks!